board/safety/safety_hyundai.h

```diff
@@ -165,38 +171,62 @@ static int hyundai_rx_hook(CANPacket_t *to_push) {
       update_sample(&torque_driver, torque_driver_new);
     }
 
-    if (hyundai_longitudinal) {
-      // ACC steering wheel buttons
-      if (addr == 1265) {
-        int button = GET_BYTE(to_push, 0) & 0x7U;
+    // ACC steering wheel buttons
+    if (addr == 1265) {
+      int cruise_button = GET_BYTE(to_push, 0) & 0x7U;
+      int main_button = GET_BIT(to_push, 3U);
+
+      // used for keeping track of user button interaction
+      update_sample(&cruise_buttons, cruise_button);
+      update_sample(&main_buttons, main_button);
 
+      if (hyundai_longitudinal) {
         // exit controls on cancel press
-        if (button == HYUNDAI_BTN_CANCEL) {
+        if (cruise_button == HYUNDAI_BTN_CANCEL) {
           controls_allowed = 0;
         }
 
         // enter controls on falling edge of resume or set
-        bool set = (button == HYUNDAI_BTN_NONE) && (cruise_button_prev == HYUNDAI_BTN_SET);
-        bool res = (button == HYUNDAI_BTN_NONE) && (cruise_button_prev == HYUNDAI_BTN_RESUME);
+        bool set = (cruise_button == HYUNDAI_BTN_NONE) && (cruise_button_prev == HYUNDAI_BTN_SET);
+        bool res = (cruise_button == HYUNDAI_BTN_NONE) && (cruise_button_prev == HYUNDAI_BTN_RESUME);
         if (set || res) {
           controls_allowed = 1;
         }
 
-        cruise_button_prev = button;
+        cruise_button_prev = cruise_button;
       }
-    } else {
-      // enter controls on rising edge of ACC, exit controls on ACC off
-      if (addr == 1057) {
-        // 2 bits: 13-14
-        int cruise_engaged = (GET_BYTES_04(to_push) >> 13) & 0x3U;
-        if (cruise_engaged && !cruise_engaged_prev) {
-          controls_allowed = 1;
+    }
+
+    // enter controls on rising edge of ACC and user button press, exit controls when ACC off
+    if (!hyundai_longitudinal && (addr == 1057)) {
+      // 2 bits: 13-14
+      int cruise_engaged = (GET_BYTES_04(to_push) >> 13) & 0x3U;
+      if (cruise_engaged && !cruise_engaged_prev) {
+        bool cruise_pressed_recent = false;
+        bool main_pressed_recent = false;
+        for (int i = 0; i < MIN(cruise_buttons.length, PREV_BUTTON_SAMPLES); i++) {
```

Comment:
Can make a function for these two common loops, but it means a global function in safety.h that's make-specific and only ever used here.